architecture/2025-04-compute-platform-extensibility.md

```diff
@@ -0,0 +1,505 @@
+param aciscope string = '/subscriptions/<subscription id/resourceGroups/<resource group name>'
+resource env 'Applications.Core/environments@2023-10-01-preview' = {
+  name: 'aci-env'
+  properties: {
+    compute: {
+      kind: 'aci'
+      resourceGroup: aciscope
+    }
+    providers: {
+      azure: {
+        scope: aciscope
+      }
+    }
+  }
+}
+```
+
+### Sample Input: Portable Resources
+
+[Portable Resources](https://docs.radapp.io/guides/author-apps/portable-resources/overview/) are expected to work on ACI with no changes from the way they would be configured for Kubernetes:
+
+```bicep
+extension radius
+@description('Specifies the environment for resources.')
+param environment string
+
+@description('Specifies the application for resources.')
+param application string
+
+resource redis 'Applications.Datastores/redisCaches@2023-10-01-preview'= {
+  name: 'myredis'
+  properties: {
+    environment: environment
+    application: application
+  }
+}
+```
+
+#### Output
+
+The above example will deploy on either Kubernetes or ACI.
+
+## Design
+
+### High Level Design
+
+The integration with ACI will be achieved by extending the existing Radius control plane, which will continue to be hosted in Kubernetes, to support ACI as a target deployment environment. The following summarizes the key changes and enhancements that will be made to the Radius codebase to support ACI deployments:
+
+1. Environment Definition and Setup: The Radius control plane will be updated to recognize and process environment definition files (env.bicep) that specify ACI as the target compute platform. This will involve adding support for ACI-specific configurations and parameters in the environment setup process.
+1. Application Definition and Deployment: The application definition files (app.bicep) will be extended to include ACI-specific properties and configurations. The Radius deployment engine will be enhanced to interpret these properties and deploy applications to ACI environments. This includes creating and managing ACI container groups, configuring networking, and handling ACI-specific resource settings.
+1. Container Definitions: The container definitions within the application definition files will allow users to deploy their containers to both Kubernetes and ACI without significant changes.
+1. Extension Properties, aka "punch-through": The Radius codebase will be enhanced to support extension properties in the application definition files. These properties will allow users to apply ACI-specific configurations to their containers, such as custom resource settings, networking configurations, and other ACI-specific parameters.
+1. Error Handling and Logging: The error handling and logging mechanisms within the Radius codebase will be updated to account for ACI-specific scenarios. This includes handling errors related to ACI resource creation, configuration, and deployment, as well as providing detailed logs for troubleshooting and diagnostics.
```

Comment:
What type of updates do you envision in error handling? I'd assume Radius would just propagate back whatever ACI returned.

Reply:
Users will need clear error messages when they have used a Radius feature that is not supported by an ACI deployment. I will update the text here to clarify that.